packages/vx-xychart/src/providers/DataProvider.tsx

```diff
@@ -0,0 +1,136 @@
+import { ScaleConfig, createScale, updateScale, ScaleConfigToD3Scale, ScaleInput } from '@vx/scale';
+import { extent as d3Extent } from 'd3-array';
+import React, { useContext, useMemo } from 'react';
+import createOrdinalScale from '@vx/scale/lib/scales/ordinal';
+import { AxisScaleOutput } from '@vx/axis';
+import { XYChartTheme } from '../types';
+import ThemeContext from '../context/ThemeContext';
+import DataContext from '../context/DataContext';
+import useDataRegistry from '../hooks/useDataRegistry';
+import useDimensions from '../hooks/useDimensions';
+
+/** Props that can be passed to initialize/update the provider config. */
+export type DataProviderProps<
+  XScaleConfig extends ScaleConfig<AxisScaleOutput>,
+  YScaleConfig extends ScaleConfig<AxisScaleOutput>
+> = {
+  theme?: XYChartTheme;
+  xScale: XScaleConfig;
+  yScale: YScaleConfig;
+  children: React.ReactNode;
+};
+
+export default function DataProvider<
+  XScaleConfig extends ScaleConfig<AxisScaleOutput>,
+  YScaleConfig extends ScaleConfig<AxisScaleOutput>,
+  Datum = unknown,
+  DataKey extends string = string
+>({
+  theme: propsTheme,
+  xScale: xScaleConfig,
+  yScale: yScaleConfig,
+  children,
+}: DataProviderProps<XScaleConfig, YScaleConfig>) {
+  type XScale = ScaleConfigToD3Scale<XScaleConfig, AxisScaleOutput>;
+  type YScale = ScaleConfigToD3Scale<YScaleConfig, AxisScaleOutput>;
+  type XScaleInput = ScaleInput<ScaleConfigToD3Scale<XScaleConfig, AxisScaleOutput>>;
+  type YScaleInput = ScaleInput<ScaleConfigToD3Scale<YScaleConfig, AxisScaleOutput>>;
+
+  const contextTheme = useContext(ThemeContext);
+  const theme = propsTheme || contextTheme;
+  const [{ width, height, margin }, setDimensions] = useDimensions();
+  const innerWidth = width - (margin?.left ?? 0) - (margin?.right ?? 0);
+  const innerHeight = height - margin?.top ?? 0 - margin?.bottom ?? 0;
+
+  const dataRegistry = useDataRegistry<XScaleConfig, YScaleConfig, Datum, DataKey>();
+
+  // keys are the memoization value that matters for this memo
+  const registryKeys = dataRegistry.keys;
+
+  // @TODO(chris) factor out to a separate hook
+  const scales = useMemo(() => {
+    const registryEntries = registryKeys.map(key => dataRegistry.get(key));
+
+    const xValues = registryEntries.reduce<XScaleInput[]>(
+      (combined, entry) =>
+        entry ? combined.concat(entry.data.map((d: Datum) => entry.xAccessor(d))) : combined,
+      [],
+    );
+    const yValues = registryEntries.reduce<YScaleInput[]>(
+      (combined, entry) =>
+        entry ? combined.concat(entry.data.map((d: Datum) => entry.yAccessor(d))) : combined,
+      [],
+    );
+
+    const xMin = margin.left;
+    const xMax = width - margin.right;
+    const yMin = margin.top;
+    const yMax = height - margin.bottom;
+
+    const { type: xType, ...restXScaleConfig } = xScaleConfig;
+    const { type: yType, ...restYScaleConfig } = yScaleConfig;
+
+    let xScale = createScale({
+      type: xType,
```

Comment:
this is failing types still, @kristw any thoughts?

![image](https://user-images.githubusercontent.com/4496521/91257209-3015ed80-e71e-11ea-8532-e2f8b9210765.png)

Reply:
What if we do another way around?

```ts
const xScale = createScale(xScaleConfig);
type XScale = typeof xScale;
type XScaleInput = ScaleInput<XScale>;
// update domain and range
xScale.domain(...).range(...);
```

Reply:
oh interesting. I get an error even when passing `xScaleConfig: XScaleConfig extends ScaleConfig<AxisScaleOutput>` directly to `createScale` 🤔 

![image](https://user-images.githubusercontent.com/4496521/91352304-8296ee80-e79e-11ea-9149-d7855f2fdb3b.png)

Reply:
it looks like it might be inferring that it is the first function overload signature (linear scale) based on `typeof xScale`

![image](https://user-images.githubusercontent.com/4496521/91353050-9131d580-e79f-11ea-9d59-1b4795deb9b9.png)

similar interpretation based on the accessor return value 
![image](https://user-images.githubusercontent.com/4496521/91353258-de15ac00-e79f-11ea-84a8-362727441a1b.png)

and re-assigning the `config.domain` later on
![image](https://user-images.githubusercontent.com/4496521/91353336-fa194d80-e79f-11ea-8bc1-98c42d8b4995.png)


and

Reply:
seems like it struggles to pick the right overload, it complains with `AnyD3Scale` too 🤔 

![image](https://user-images.githubusercontent.com/4496521/91353754-904d7380-e7a0-11ea-8b2e-3f8866c8a100.png)

Reply:
Hmm, I might have to add another overload at the end to catch all.

Reply:
Try adding this to `createScale.ts` (line 110, above the comment `// Actual implementation)

```ts
function createScale<
  Output = DefaultOutput,
  DiscreteInput extends StringLike = StringLike,
  ThresholdInput extends DefaultThresholdInput = DefaultThresholdInput
>(
  config: ScaleConfig<Output, DiscreteInput, ThresholdInput>,
): D3Scale<Output, DiscreteInput, ThresholdInput>;

// Actual implementation
```

Reply:
^I thiiiink I got this to work, but only if that's the _first_ overload, not the last. I'll try to see if the other overloads work as expected still (e.g., linear config => linear scale)

Reply:
hmm. usually the broader overload should be the last one. i'll try to see if there are ways to improve this. agree it is not ideak,

Reply:
I updated this to be the first overload for now